functions-python/gbfs_validator/src/gbfs_utils.py

```diff
@@ -0,0 +1,155 @@
+from dataset_service.main import (
+    Status,
+)
+
+VALIDATOR_URL = os.getenv(
+    "VALIDATOR_URL",
+    "https://gbfs-validator.mobilitydata.org/.netlify/functions/validator-summary",
+)
+
+
+def fetch_gbfs_files(url: str) -> Dict[str, Any]:
+    """Fetch the GBFS files from the autodiscovery URL."""
+    response = requests.get(url)
+    response.raise_for_status()
+    return response.json()
+
+
+def upload_gbfs_file_to_bucket(
+    bucket: storage.Bucket, file_url: str, destination_blob_name: str
+) -> str:
+    """Upload a GBFS file to a Cloud Storage bucket."""
+    response = requests.get(file_url)
+    response.raise_for_status()
+    blob = bucket.blob(destination_blob_name)
+    blob.upload_from_string(response.content)
+    blob.make_public()
+    logging.info(f"Uploaded {destination_blob_name} to {bucket.name}.")
+    return blob.public_url
+
+
+def create_gbfs_json_with_bucket_paths(
+    bucket: storage.Bucket, gbfs_data: Dict[str, Any], stable_id: str
+) -> str:
+    """
+    Create a new gbfs.json with paths pointing to Cloud Storage and upload it.
+    @param bucket: The Cloud Storage bucket.
+    @param gbfs_data: The GBFS data.
+    @param stable_id: The stable ID of the feed.
+    @return: The public URL of the new gbfs.json.
+    """
+    new_gbfs_data = gbfs_data.copy()
+    today = datetime.now().strftime("%Y-%m-%d")
+
+    for feed_key, feed in new_gbfs_data["data"].items():
+        if isinstance(feed["feeds"], dict):
+            for feed_language, feed_info in feed["feeds"].items():
+                old_url = feed_info["url"]
+                blob_name = f"{stable_id}/{stable_id}-{today}/{feed_info['name']}_{feed_language}.json"
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        elif isinstance(feed["feeds"], list):
+            for feed_info in feed["feeds"]:
+                old_url = feed_info["url"]
+                blob_name = f"{stable_id}/{stable_id}-{today}/{feed_info['name']}.json"
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        else:
+            logging.warning(f"Unexpected format in feed: {feed_key}")
+
+    # Save the new gbfs.json in the bucket
+    new_gbfs_data["last_updated"] = today
+    new_gbfs_blob = bucket.blob(f"{stable_id}/{stable_id}-{today}/gbfs.json")
+    new_gbfs_blob.upload_from_string(
+        json.dumps(new_gbfs_data), content_type="application/json"
+    )
+    new_gbfs_blob.make_public()
+    return new_gbfs_blob.public_url
+
+
+def save_trace_with_error(trace, error, trace_service):
+    """Helper function to save trace with an error."""
+    trace.error_message = error
+    trace.status = Status.FAILED
+    trace_service.save(trace)
+
+
+def create_snapshot(stable_id: str, feed_id: str, hosted_url: str) -> Gbfssnapshot:
+    """Create a new Gbfssnapshot object."""
+    today = datetime.now().strftime("%Y-%m-%d")
```

Comment:
I have seen this multiple times across the code. All points to the creation of a snapshot ID. We should centralize the creation of the ID to make sure if the ID pattern changes, there is only one place to change.

Reply:
good catch! i have created a `get_snapshot_id` function that is called from all function that constructed the id